proposals/p1885.md

```diff
@@ -0,0 +1,735 @@
+
+-   ranges `for n in 1..100`
+-   iterators `for element in container.iter()` or `.iter_mut()`
+
+The `Iterator` type can be used with custom types using the
+[following syntax](https://doc.rust-lang.org/rust-by-example/trait/iter.html)
+(though specific):
+
+```rust
+struct Fibonacci {
+    curr: u32,
+    next: u32,
+}
+
+impl Iterator for Fibonacci {
+    type Item = u32;
+    fn next(&mut self) -> Option<Self::Item> {
+        // calculate and yield value
+        [...]
+    }
+}
+
+for i in fibonacci().take(4) {
+    ...
+}
+```
+
+#### Typescript
+
+Typescript uses the
+[`.iterator` property](https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html#iterables),
+along with
+[two forms](https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html#forof-statements)
+of range-based for loops: `for..in` and `for..of`.
+
+```ts
+let pets = new Set(['Cat', 'Dog', 'Hamster']);
+pets['species'] = 'mammals';
+for (let pet in pets) {
+    console.log(pet); // "species"
+}
+for (let pet of pets) {
+    console.log(pet); // "Cat", "Dog", "Hamster"
+}
+```
+
+Using the `.iterator` property, Typescript allow to lazily generate data easily
+using functions, classes, or objects:
+
+```ts
+const reverse = (arr) => ({
+    [Symbol.iterator]() {
+        let i = arr.length;
+        return {
+            next: () => ({
+                value: arr[--i],
+                done: i < 0,
+            }),
+        };
+    },
+});
+```
+
+#### Go
+
+Go does not officially support customizing `for` for user types.
+
+## Proposal
+
+Provide interfaces that can be implemented by user types to enable support for
+ranged-for loops.
+
+We propose:
+
+-   A new `Iterate` interface to support for loops
+-   Using a cursor-based approach
+-   Making the `for` loop value `let` by default, that is, non-reassignable
+
+Below is a high-level example of this concept:
+
+```carbon
+// Add support for `for`
+class MyRange {
+  impl as Iterate where .ElementType = i32 and .CursorType = i32 {
+    // Implement `Iterate`
+  }
+}
+// Usage
+let my_range: MyRange = {...};
+for (item: auto in my_range) {
+  // Do something with `item`
+}
+```
+
+## Details
+
+### `for` with immutable values
+
+#### `Iterate` interface
+
+This proposal exposes a new `Iterate`. This interface:
+
+-   Relies on a cursor-based approach: minimize implementation effort for
+    developers, could support R-values.
+-   Expects an `ElementType` and `CursorType`.
+-   Combines "advance", "get", and "bounds check" into a single `Next(cursor)`
+    method that returns an optional value.
+
+The `Iterate` interface is:
+
+```carbon
+interface Iterate {
+  let ElementType:! type;
+  let CursorType:! type;
+  fn NewCursor[self: Self]() -> CursorType;
+  fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType);
+}
+```
+
+A naive Carbon implementation of the for loop could be:
+
+```carbon
+let cursor: range.(Iterate.CursorType) = range.(Iterate.NewCursor)();
```

Comment:
if the curser is mutated during the iteration, shouldn't it be `var curser : ...`?

Reply:
You are correct. Good catch!

```suggestion
var cursor: range.(Iterate.CursorType) = range.(Iterate.NewCursor)();
```

Reply:
Thanks!